src/region.c

```diff
@@ -53,25 +53,144 @@ gdip_region_new ()
+	if (r1->Y == r2->Y && r1->X == r2->X)
+		return 0;
+	if (r1->Y > r2->Y || (r1->Y == r2->Y && r1->X > r2->X))
+		return 1;
+	return -1;
+}
+
+static void
+gdip_sort_rect_array (GpRectF* array, int length) {
+	qsort (array, length, sizeof (GpRectF), gdip_compare_rectf);
+}
+
+// Not a mistake in the name, it is for re-sorting nearly-sorted data.
+// Insertion sort.
+static void
+gdip_sort_rect_array_sorted (GpRectF* array, int length) {
+	GpRectF rect;
+	GpRectF *i, *j;
+
+	for (i = array + 1; i < array + length; i++) {
+		rect = *i;
+		for (j = i - 1; j >= array && gdip_compare_rectf (j, &rect) > 0; j--) {
+			*(j + 1) = *j;
+		}
+		*(j + 1) = rect;
+	}
+}
+
 static GpStatus
-gdip_add_rect_to_array (GpRectF** srcarray, int* elements, const GpRectF* rect)
-{
-	GpRectF *array, *next;
+gdip_extend_rect_array (GpRectF** srcarray, int* elements, int* capacity) {
+	GpRectF *array;
+	int newCapacity = -1;
+
+	if (capacity) {
+		if (*srcarray == NULL) {
+			if (*capacity < 1)
+				*capacity = 5; // starting capacity if we're given a size of zero
+			newCapacity = *capacity;
+		} else if (*elements == *capacity) {
+			newCapacity = *elements * 2;
+		}
+	} else {
+		newCapacity = *elements + 1;
+	}
 
-	array = GdipAlloc (sizeof (GpRectF) * (*elements + 1));
+	if (newCapacity > 0) {
+		array = GdipAlloc (sizeof (GpRectF) * newCapacity);
+		if (!array)
+			return OutOfMemory;
+
+		memcpy (array, *srcarray, sizeof (GpRectF) * (*elements));
+
+		if (*srcarray)
+			GdipFree (*srcarray);
+
+		*srcarray = array;
+		if (capacity)
+			*capacity = newCapacity;
+	}
+	return Ok;
+}
+
+static GpStatus
+gdip_trim_rect_array (GpRectF** srcarray, int elements) {
+	GpRectF *array;
+
+	array = GdipAlloc (sizeof (GpRectF) * elements);
 	if (!array)
 		return OutOfMemory;
 
-	memcpy (array, *srcarray, sizeof (GpRectF) * (*elements));
+	memcpy (array, *srcarray, sizeof (GpRectF) * elements);
 
 	if (*srcarray)
 		GdipFree (*srcarray);
 
-	next = array;
+	*srcarray = array;
+	return Ok;
+}
+
+static GpStatus
+gdip_add_rect_to_array (GpRectF** srcarray, int* elements, int* capacity, const GpRectF* rect)
+{
+	GpRectF *next;
+	GpStatus status;
+
+	status = gdip_extend_rect_array (srcarray, elements, capacity);
+	if (status != Ok)
+		return status;
+
+	next = *srcarray;
 	next += (*elements);
 	memcpy (next, rect, sizeof (GpRectF));
 
-	*srcarray = array;
+	*elements = *elements + 1;
+
+	return Ok;
+}
+
+static GpRectF*
+gdip_binsearch_rect_array (GpRectF* array, int elements, const GpRectF* search, int* index)
+{
+	GpRectF *next;
+	int upper = elements, lower = 0, mid;
+
+	while (upper > lower) {
+		mid = (upper + lower) / 2;
+		next = array + mid;
+		if (gdip_compare_rectf (search, next) > 0) {
+			lower = mid + 1;
+		} else {
+			upper = mid;
+		}
+	}
+	next = array + lower;
+	if (index)
+		*index = lower;
+	return next;
+}
+
+static GpStatus
+gdip_add_rect_to_array_sorted (GpRectF** srcarray, int* elements, int* capacity, const GpRectF* rect)
+{
```

Comment:
This function is not used as far as I can tell. It would certainly make sense to use it at the place where `gdip_sort_rect_array_sorted` is used now...

Reply:
Correct. I was using it there initially, but changed to using the sort and didn't remove the method. In union the modifications to `recttrg` require sorting after them, so I decided it would probably be better to do one sort that can handle both of those. However, on thinking about it again, only one of the two ever happens (either `rslt` is empty, or the remnant of `recttrg` will be empty). Therefore it would be better to use this there as you suggest.